### PR TITLE
Switch to ruff

### LINKED ISSUE #429: Use ruff for code formatting/linting

### PR DESCRIPTION
**Pull request by alifbe**
- Use ruff for code linting and formatting close #429 
- Update contributing documentation
- Remove sphinx pin

### CODE REVIEW

**Review by mferrera — COMMENTED** [review]
I'm a bit concerned about performance implications of remove `inplace=True` from the dataframes. Maybe @berland should comment about this (he may also object to removing pylint stuff)

File: pyscal/factory.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ class PyscalFactory(object):
     @staticmethod
     def create_water_oil(
         params: Optional[Dict[str, float]] = None, fast: bool = False
-    ) -> WaterOil:
+    ) -> WaterOil:  # noqa
```
[the remark this answers is not in the document]
Comment (alifbe) [reply pyscal/factory.py:157]: Good catch. Thanks. I was trying to ignore C901 here but it didn't work. So, I turned off C901 in ruff

File: pyscal/gasoil.py [hunk 2]
```diff
@@ -154,7 +154,7 @@ def __init__(
         sg_list.sort()
         self.table = pd.DataFrame(sg_list, columns=["SG"])
         self.table["sgint"] = list(map(round, self.table["SG"] * SWINTEGERS))
-        self.table.drop_duplicates("sgint", inplace=True)
+        self.table = self.table.drop_duplicates("sgint")
```
[the remark this answers is not in the document]
Comment (alifbe) [reply pyscal/gasoil.py:157]: It's `PD` option in ruff that trigger this change. I saw in pandas documentation that inplace parameter will/is deprecated but cannot find it today. But here is the issue that I found today.
https://github.com/pandas-dev/pandas/issues/16529

File: pyscal/factory.py [hunk 3]
```diff
@@ -1218,11 +1208,8 @@ def filter_nan_from_dict(params: dict) -> dict:
     """
     cleaned_params = {}
     for key, value in params.items():
-        if isinstance(value, str):
+        if (isinstance(value, str)) or (not np.isnan(value)):
```
[the remark this answers is not in the document]
Comment (alifbe) [reply pyscal/factory.py:1211]: That's true :+1:

**Review by mferrera — APPROVED** [review]

**Comment by alifbe** [comment]
> I'm a bit concerned about performance implications of removing `inplace=True` from the dataframes. Maybe @berland should comment about this (he may also object to removing pylint stuff)

I didn't know that `ruff` doesn't replace `pylint` :blush: 

It seems like there are quite a few issues in main branch found by `pylint`. Instead of putting it back here, I am thinking to solve it in another PR. Created issue #431

**Comment by mferrera** [comment]
For what it's worth, I think pylint is typically used to get a sense of some code quality or improvements rather than making sure code satisfies all of its conditions. You'll almost never see pylint included in linting CI workflows for that reason. It's a good barometer but not necessarily a bar to meet.

**Pull request merged**